### PR TITLE
:sparkles: Test automation FMC Edit Shipping Details

### PR DESCRIPTION
**Pull request by arianenld**
> ACCEPTANCE CRITERIA #2 Edit Shipping Details

GIVEN User is logged in to storefront/<locale>
WHEN User clicks Account icon
AND clicks EDIT on Default Shipping Address
THEN an editable form group should be displayed.

### CODE REVIEW

File: storefront/Browser.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    def quit(self):
+        self.browser.quit()
+
+
+''' 
+COUNTER METHODS
+'''
+def getTestAccountCounter():
+    file = open("index.txt", "r")
```
Comment (tradrianalejandro) [inline storefront/Browser.py]: try using a more descriptive filename for the benefit of new or other developers. :)

File: storefront/Browser.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    contents = file.readline()
+    file.close()
+    return contents
+
+def setTestAccountCounter(ctr):
```
Comment (tradrianalejandro) [inline storefront/Browser.py]: minor comment on the function names of both the getter and the setter. can we try to come up with a more intuitive name for these functions?

File: storefront/Browser.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    file = open("index.txt", "w")
```
Comment (tradrianalejandro) [inline storefront/Browser.py]: Curious. What could happen if the file gets corrupted or becomes non-existent in between its opening and closing?

File: storefront/globals.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+CHROME = "drivers/chomedriver"
```
Comment (tradrianalejandro) [inline storefront/globals.py]: good practice to use all caps when declaring constants. We could add and proly agree on using underscores when separating words, e.g. US_LOCALE.

File: storefront/FMC-AC-2.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+from globals import *
+from Browser import *
+import time
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: aesthetics only ;) 
please move import time to top so we can have some sorta groupings.

File: storefront/FMC-AC-2.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    editTester.fillUpEditAddressFormFields()
+    editTester.wait(3)
+
+    editTester.quit()
+
+    print("automation success")
+
+# FOR US
+editFunctionShouldBeFunctinonal(USLOCALE, webdriver.Chrome("../drivers/chromedriver"))
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: we want function names to be as descriptive and as intuitive as possible. :)

File: storefront/FMC-AC-2.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        self.findByName(lastName).clear()
+        self.typeInElement(self.findByName(lastName), "Nulud")
+        self.typeInElement(self.findByName(primaryPhone), "09650552465")
+        self.typeInElement(self.findByName(companyPhone), "09650552138")
+        self.typeInElement(self.findByName(streetAddress), "466 Baging Silang Cutcut Angeles City")
+        self.typeInElement(self.findByName(addressLine2), "Plantersville Mabiga Mabalacat City")
+        self.typeInElement(self.findByName(city), "Angeles")
+        self.typeInElement(self.findByName(postcode), "2009")
+
+def editFunctionShouldBeFunctinonal(locale, browser):
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: we'll proly be running our tests using chrome most of the time. default parameters will be beneficial here.

File: storefront/FMC-AC-2.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+    editTester = EditFunctionTest(browser)
+    editTester.visitUrl(locale)
+    editTester.assertURL(locale)
+    print("url check")
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: kill this ;)

File: storefront/FMC-AC-2.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        self.findByClassName("ht-account").click()
+
+        self.wait(3)
+
+        # edit icon
+        self.findByClassName("edit-address").click()
+
+        # shipping address details page
+        self.findByName(firstName).clear()
+        self.typeInElement(self.findByName(firstName), "Ariane")
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: create a struct-style or whatever test data that we could pull from as default information. Use valid information.

File: storefront/globals.py (hunk 4 above)
Comment (arianenld) [reply storefront/globals.py]: done

File: storefront/FMC-AC-2.py (hunk 5 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 6 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 8 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 9 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 7 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 5 above)
Comment (trzaeraldlungos) [reply storefront/FMC-AC-2.py]: I would like to also add that we should consider adding some python conventions so the formatting and positions of the imports are standardized.

File: storefront/FMC-AC-2.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
+from globals import *
+from Browser import *
+import time
+from selenium import webdriver
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: remove this as already imported on Browser class. :)

File: storefront/FMC-AC-2.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
+from globals import *
+from Browser import *
+from selenium.webdriver.support.select import Select
+
+class EditFunctionTest(Browser):
+    def getHeaderLogo(self):
+        return self.browser.find_element_by_class_name("header-logo")
+
+    def trimURL(self):
+        return self.browser.current_url[0:28]
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: make this more generic to handle URLs across different environments (stage, prodtest, prod)

File: storefront/FMC-AC-2.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+
+    def checkMyAccountButtonElement(self):
+        try:
+            assert self.findByClassName("ht-account")
+            print("Account Icon Element complete")
+        except:
+            print("Account Icon Element missing")
+
+
+    def checkAccountMaintenanceElements(self):
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: include visibility

File: storefront/FMC-AC-2.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        # account icon
+        self.findByClassName("ht-account").click()
+
+        self.checkAccountMaintenanceElements()
+
+        # edit icon
+        self.findByClassName("edit-address").click()
+
+        # shipping address details page
+        print(self.trimURL())
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: getLocale()

File: storefront/FMC-AC-2.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+        if (self.trimURL() == AU_LOCALE):
+            self.findByName(firstName).clear()
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: refactor clearing of elements to resetForm() or the like

File: storefront/globals.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+lastName = "lastName"
+primaryPhone = "primaryPhone"
+company = "company"
+streetAddress = "streetAddress"
+addressLine2 = "addressLine2"
+city = "subdivision"
+stateProvince = "state"
+postcode = "zipCode"
+
+#US EDIT SHIPPING DETAILS VALUE
```
Comment (tradrianalejandro) [inline storefront/globals.py]: use convention for constants, e.g. US_CITY

Comment (tradrianalejandro) [inline storefront/globals.py]: use dictionary for personal details

File: storefront/FMC-AC-2.py [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+            self.findByName(city).clear()
+            self.typeInElement(self.findByName(city), usCity)
+
+            Select(self.findByID(stateProvince)).select_by_index(0)
+            Select(self.findByID(stateProvince)).select_by_index(1)
+
+            self.findByName(postcode).clear()
+            self.typeInElement(self.findByName(postcode), usPostcode)
+
+        else:
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: invalid locale

File: storefront/FMC-AC-2.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+            self.findByName(streetAddress).clear()
+            self.typeInElement(self.findByName(streetAddress), auStreetAddress)
+
+            self.findByName(addressLine2).clear()
+            self.typeInElement(self.findByName(addressLine2), auAddressLine2)
+
+            self.findByName(city).clear()
+            self.typeInElement(self.findByName(city), auCity)
+
+            Select(self.findByID(stateProvince)).select_by_index(0)
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: proly better when strings are selected via code so we can avoid magic numbers

File: storefront/FMC-AC-2.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+            self.typeInElement(self.findByName(addressLine2), usAddressLine2)
+
+            self.findByName(city).clear()
+            self.typeInElement(self.findByName(city), usCity)
+
+            Select(self.findByID(stateProvince)).select_by_index(0)
+            Select(self.findByID(stateProvince)).select_by_index(1)
+
+            self.findByName(postcode).clear()
+            self.typeInElement(self.findByName(postcode), usPostcode)
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: 👍 on consistency in camel case convention

File: storefront/FMC-AC-2.py [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+        try:
+            assert self.findByID("desktop-login-username")
+            assert self.findByID("desktop-login-password")
+            assert self.findByID("desktop-user-login-btn")
+            assert self.findByClassName("dh-forgot-pass")
+            print("Sign in Modal Elements complete")
+        except:
+            print("Sign in Modal Elements incomplete")
+
+    def checkHomePageLogo(self):
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: in general, change "check" to "verify"

File: storefront/FMC-AC-2.py (hunk 14 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 13 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/globals.py (hunk 15 above)
Comment (arianenld) [reply storefront/globals.py]: done

File: storefront/FMC-AC-2.py (hunk 16 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 17 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 19 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py [hunk 20]
```diff
@@ -6,187 +6,139 @@ class EditFunctionTest(Browser):
...
-            Select(self.findByID(stateProvince)).select_by_index(0)
-            Select(self.findByID(stateProvince)).select_by_index(1)
-
-            self.findByName(postcode).clear()
-            self.typeInElement(self.findByName(postcode), usPostcode)
-
+        if (self.getLocale() == AU_LOCALE):
+            self.resetEditAddressFormFields()
+            for element, value in AU_USER_SHIPPING_DETAILS.items():
+                if element == "state":
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: add VIC to test data to simplify code

File: storefront/FMC-AC-2.py [hunk 21]
```diff
@@ -6,187 +6,139 @@ class EditFunctionTest(Browser):
...
+                   self.typeInElement(self.findByID("state"), "VIC")
+                else:
+                    self.typeInElement(self.findByName(element), value)
+
+        elif (self.getLocale() == US_LOCALE):
+            self.resetEditAddressFormFields()
+            for element, value in US_USER_SHIPPING_DETAILS.items():
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: refactor this as used twice already :)

File: storefront/FMC-AC-2.py [hunk 22]
```diff
@@ -6,187 +6,139 @@ class EditFunctionTest(Browser):
...
-
-            self.findByName(city).clear()
-            self.typeInElement(self.findByName(city), "User not registered")
-
-            Select(self.findByID(stateProvince).select_by_index(0))
-
-            self.findByName(postcode).clear()
-            self.typeInElement(self.findByName(postcode), "User not registered")
+            self.resetEditAddressFormFields()
+            for element, value in AU_USER_SHIPPING_DETAILS.items():
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: no need to fill in details. you can log this into console instead.

File: storefront/FMC-AC-2.py [hunk 23]
```diff
@@ -6,187 +6,139 @@ class EditFunctionTest(Browser):
...
-                print("Ok button missing.")
+                print("[WARNING]: OK button missing.")
         except:
-            print("Success Modal for Editing Elements incomplete")
+            print("[WARNING]: Success Modal for Editing Elements incomplete")
+
+    def verifyPageRedirectionAfterEditing(self):
+        try:
+            if self.getLocale() == US_LOCALE:
+                assert self.getCurrentURL()
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: if self.getLocale() == US_LOCALE or AU_LOCALE:
   assert
   print

File: storefront/FMC-AC-2.py (hunk 20 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 21 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 11 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: doneee :))))

File: storefront/FMC-AC-2.py (hunk 22 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 23 above)
Comment (arianenld) [reply storefront/FMC-AC-2.py]: done :)

File: storefront/FMC-AC-2.py [hunk 24]
```diff
@@ -1,13 +1,22 @@
...
 
+BUTTON_OK_XPATH = "/html/body/div[7]/div/div/div[2]/button"
+
 class EditFunctionTest(Browser):
     def getHeaderLogo(self):
         return self.browser.find_element_by_class_name("header-logo")
 
-    def getLocale(self):
-        return self.browser.current_url[0:28]
+    def getLocale(self, locale):
```
Comment (tradrianalejandro) [inline storefront/FMC-AC-2.py]: isLocaleValid or isLocaleRetrieved or sth like this

Comment (arianenld) [reply storefront/FMC-AC-2.py]: done

File: storefront/FMC-AC-2.py (hunk 5 above)
Comment (tradrianalejandro) [reply storefront/FMC-AC-2.py]: you're right zae. thanks. we can implement standardized conventions as we go in the next pull requests

**Pull request merged**